src/ducc0/fft/fft.h

```diff
@@ -178,8 +178,78 @@ struct util // hack to avoid duplicate symbols
     return std::max(size_t(1), std::min(parallel, max_threads));
     }
 #endif
+  static size_t chunk_size(size_t nwork, size_t nthreads, size_t vlen)
+    {
+    size_t chunk_size = (nwork + nthreads - 1) / nthreads;
+#ifndef DUCC0_NO_SIMD
+    // Round up to multiple of SIMD vector length.
+    chunk_size = ((chunk_size + vlen - 1) / vlen) * vlen;
+#endif
+    return chunk_size;
+    }
+  };
+
+/// Parallel FFT execution functor.
+using ThreadingFunctor = std::function<
+  void(size_t/*nthreads*/, const fmav_info &/*info*/, size_t/*axis*/,
+    size_t/*vlen*/, std::function<void(Scheduler&/*sched*/)>/*func*/)>;
+
+/// Executes FFT in a single thread.
+struct SingleThreading
+  { 
+  inline void operator()(size_t /*nthreads*/, const fmav_info &info, size_t axis,
+    size_t /*vlen*/, std::function<void(Scheduler&)> func) const
+    {
+    size_t nwork = info.size() / info.shape(axis);
+    execSingle(nwork, std::move(func));
+    }
+  };
+
+/// Executes FFT with fixed-sized chunks statically assigned to threads.
+struct StaticThreading
+  { 
+  inline void operator()(size_t nthreads, const fmav_info &info, size_t axis,
+    size_t vlen, std::function<void(Scheduler&)> func) const
+    {
+    nthreads = util::thread_count(nthreads, info, axis, vlen);
+    size_t nwork = info.size() / info.shape(axis);
+    size_t chunk_size = util::chunk_size(nwork, nthreads, vlen);
+    execStatic(nwork, nthreads, chunk_size, std::move(func));
+    }
   };
 
+// Executes FFT with fixed-sized chunks dynamically allocated to threads.
+struct DynamicThreading
```

Comment:
Dynamic/GuidedThreading are just for demonstration.  They are just plug-and-play to replace the underlying threading routine.

Reply:
I like the idea! Still, for things like DynamicThreading and GuidedThreading to really be useful, they must be supplied with a desired (initial) chunk size; StaticThreading will work fine as it is, but still could use an optional chunk size parameter in the constructor, in case the users want finer control over how the work items are processed.

As things are written at the moment, I think that StaticThreading and DynamicThreading will behave in exactly the same way in most situations, as there is only one work item per thread,and each thread will therefore get exactly one item to process. (Of course, if some threads are completely unresponsive in the DynamicThreading case, all the work will be done by the rest of the threads, but that's probably not the typical use case.)

So we would need a heuristic for a good chunk size to pick in the case that dynamic scheduling is used. But this is something we can worry about when it will be actually used!

Reply:
> the chunk size is computed internally below (line 229). I just split up the work evenly into nthreads and rounded up to the nearest multiple of vlen.

I'm aware of this, but for dynamic scheduling this choice is not really helpful; if there is exactly one chunk of work for every thread, there can be no dynamic load balancing.
`ExecDynamic` should be used in sitations where each work item takes an unpredictable amount of time, and every thread just grabs a new chunk when it has finished with the current one. So in this case typically `n_chunk>>n_threads`.

Guided scheduling would take a maximum anda minimum allowed chunk size, making things even more flexible (and complicated).

Reply:
I see, I assumed `Dynamic` was for the case of threads being busy with some other workload (which they often are in TF, used for computing other ops in parallel).

So maybe the inputs to the `operator()` itself should be
- nwork
- nthreads (_suggested_ number of threads, which may or may not be the true number of threads used)
- chunksize (_suggested_ chunk size for dividing up work across the suggested nthread)

This would be more in-line with the corresponding functions in `threading.h`, and would apply to both the N-D and 1-D cases.  In TF I'll adjust our internal scheduler to recursively divide until we hit the suggested `chunksize`.

The `DynamicThreading` constructor can take an optional argument to control a maximum chunk size.  Or we can remove this one entirely - the only one needed is `StaticThreading`.  I just put the other ones in for demonstration.